File: a3/main.py
with open("input.txt", "r") as f:
	rucksacks = [line.strip() for line in f.readlines()]

def calc_priority(a):
	a = ord(a)
	if ord('a') <= a <= ord('z'):
		return a - ord('a') + 1
	if ord('A') <= a <= ord('Z'):
		return a - ord('A') + 27

def intersect(a, b):
	found = []
	for item in a:
		if item in b:
			found.append(item)
	return found

def calc_similar(a, b):
	return calc_priority(intersect(a, b)[0])

def solve1(rucksacks):
	total = 0
	for rucksack in rucksacks:
		total += calc_similar(rucksack[len(rucksack)//2:], rucksack[:len(rucksack)//2])
	return total

def solve2(rucksacks):
	total = 0
	for i in range(len(rucksacks) // 3):
		first, second, third = tuple(rucksacks[i*3:i*3+3])
		total += calc_similar(intersect(first, second), third)
	return total

print(solve1(rucksacks))
print(solve2(rucksacks))

File: a3/main.rs
use std::fs;

fn calc_priority(a: char) -> i32 {
	if 'a' <= a && a <= 'z' {
		return a as i32 - 'a' as i32 + 1 as i32;
	}
	if 'A' <= a && a <= 'Z' {
		return a as i32 - 'A' as i32 + 27 as i32;
	}
	unreachable!();
}

fn intersect(a: &String, b: &String) -> String {
	a.chars().filter(|&c| b.contains(c)).collect()
}

fn calc_similar(a: &String, b: &String) -> i32 {
	return calc_priority(intersect(a, b).chars().next().unwrap())
}

fn solve1(rucksacks: &[String]) -> i32 {
	rucksacks.iter().fold(0, |a, l| {
		let (first, second) = l.split_at(l.chars().count() / 2);
		a + calc_similar(&first.to_string(), &second.to_string())
	})
}

fn solve2(rucksacks: &[String]) -> i32 {
	(0..(rucksacks.len() / 3)).fold(0, |t, i| {
		let (first, second, third) = if let [first, second, third] = &rucksacks[(i*3)..(i*3+3)] {
			(first, second, third)
		} else {
			unreachable!()
		};
		t + calc_similar(&intersect(first, second), third)
	})
}

fn main() {
	let rucksacks: Vec<String> = fs::read_to_string("input.txt")
		.expect("File does not exist")
		.split("\n")
		.map(|a| a.to_string())
		.collect();
	println!("{:?}", solve1(&rucksacks));
	println!("{:?}", solve2(&rucksacks));
}
